contracts/libraries/LockDataLibrary.sol

```diff
@@ -2,49 +2,91 @@
+///      that tracks current lockers. The LOCK_DATA storage slot for this data structure,
+///      passed in as a LockData custom type that is a uint256, stores not just the current
 ///      length of the queue but also the global count of non-zero deltas across all lockers.
-///      The values of the data structure start at OFFSET, and each value is a locker address.
+///      The values of the data structure start at LOCKERS, and each value is a locker address.
 library LockDataLibrary {
-    uint256 private constant OFFSET = uint256(keccak256("LockData"));
+    using LockDataLibrary for LockData;
 
-    /// @dev Pushes a locker onto the end of the queue, and updates the sentinel storage slot.
-    function push(IPoolManager.LockData storage self, address locker) internal {
-        // read current value from the sentinel storage slot
-        uint128 length = self.length;
+    uint256 private constant LOCK_DATA = uint256(keccak256("LockData"));
+    uint256 private constant LOCKERS = uint256(keccak256("Lockers"));
+
+    /// @dev Pushes a locker onto the end of the queue at the LOCKERS + length storage slot, and updates the length at the LOCK_DATA storage slot.
+    function push(LockData lockData, address locker) internal {
+        // read current length from the LOCK_DATA storage slot
+        uint128 _length = lockData.length();
         unchecked {
-            uint256 indexToWrite = OFFSET + length; // not in assembly because OFFSET is in the library scope
+            uint256 lockerSlot = LOCKERS + _length; // not in assembly because LOCKERS is in the library scope
+
             /// @solidity memory-safe-assembly
             assembly {
                 // in the next storage slot, write the locker
-                sstore(indexToWrite, locker)
+                sstore(lockerSlot, locker)
             }
-            // update the sentinel storage slot
-            self.length = length + 1;
+            lockData.update(_length + 1, lockData.nonzeroDeltaCount());
         }
     }
 
     /// @dev Pops a locker off the end of the queue. Note that no storage gets cleared.
-    function pop(IPoolManager.LockData storage self) internal {
+    function pop(LockData lockData) internal {
         unchecked {
-            self.length--;
+            lockData.update(lockData.length() - 1, lockData.nonzeroDeltaCount());
+        }
+    }
+
+    function length(LockData lockData) internal pure returns (uint128 _length) {
+        /// @solidity memory-safe-assembly
+        assembly {
+            _length := shr(128, lockData)
+        }
+    }
+
+    function nonzeroDeltaCount(LockData lockData) internal pure returns (uint128 _nonzeroDeltaCount) {
+        /// @solidity memory-safe-assembly
+        assembly {
+            _nonzeroDeltaCount := lockData
+        }
+    }
+
+    function update(LockData lockData, uint128 _length, uint128 _nonzeroDeltaCount) internal {
```

Comment:
for readability/less footguns, wonder if we can break this into 3 functions:
`lockData.incrementNonzeroDeltaCount()`
`lockData.decrementNonzeroDeltaCount()`
`lockData.delete()`

probably at the expense of some bytecode savings, but I think it would make a lot of the code nicer to look at

Reply:
Ok i added these functions but they all use similar code paths.. which I separated into toLockData and _update which does the final write